### PR TITLE
Integrate group and rollout

### PR DESCRIPTION
**Pull request by duanxianbo**
[(task:19154)](https://staging.teamyapp.com/teams/107/tasks/19154)

### CODE REVIEW

File: src/storage/graph/activator.node.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import { MaxViewersActivator, PercentageActivator, TimeRangeActivator } from "@core/entity/rollout";
+import { LocalStore } from "../syncer/localStore";
+import { ActivatorState, MaxViewersActivatorState, PercentageActivatorState, TimeRangeActivatorState } from "../states/activator.state";
+
+
```
Comment (naturetime10) [inline src/storage/graph/activator.node.ts]:
```suggestion
```

File: src/storage/graph/app.node.ts [hunk 2]
```diff
@@ -6,6 +6,11 @@ import { AppInstallationNode } from './appInstallation.node';
 import { AppSecretNode } from './appSecret.node';
 import { AppVersionNode } from './appVersion.node';
 import { TeamNode } from './team.node';
+import { StaticGroup } from '@core/entity/group';
+import { FilterGroupNode, StaticGroupNode } from './group.node';
+import { RolloutNode } from './rollout.node';
+import { UserNode } from './user.node';
+
```
Comment (naturetime10) [inline src/storage/graph/app.node.ts:14]:
```suggestion
```

File: src/storage/graph/app.node.ts [hunk 3]
```diff
@@ -54,6 +59,66 @@ export class AppNode implements App {
...
+                        appState.userGroupRelations, 
+                        (userId: number) => new UserNode(this.localStore, appState.users[userId])
+                    );
+                case 'FILTER':
+                    const filterGroup = appState.filterGroups[groupId];
+                    return new FilterGroupNode(this.localStore, group, filterGroup);
+            }
+        });
+    }
+
```
Comment (naturetime10) [inline src/storage/graph/app.node.ts:90]:
```suggestion
```

File: src/storage/graph/group.node.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+            return this.group.rolloutIds.map((rolloutId) => {
+                return new RolloutNode(this.localStore, appState.rollouts[rolloutId]);
+            })
+        }
+
+        public get app(): AppNode {
+            return new AppNode(this.localStore, this.localStore.getState().apps[this.group.appId]);
+        }
+}
+
```
Comment (naturetime10) [inline src/storage/graph/group.node.ts:64]:
```suggestion
```

File: src/storage/states/app.state.ts [hunk 5]
```diff
@@ -18,13 +18,29 @@ import { AppSecretState } from './appSecret.state';
 import { TaskLinkState } from './taskLink.state';
 import { TeamAppVersionState } from './teamAppVersion.state';
 import { TagState } from './tag.state';
+import { FilterGroupState, GroupState } from './group.state';
+import { ActivatorState, MaxViewersActivatorState, PercentageActivatorState, TimeRangeActivatorState } from './activator.state';
+import { ExperimentVersionSelectorState, StaticVersionSelectorState, VersionSelectorState } from './versionSelector.state';
+import { RolloutState } from './rollout.state';
+import { MemberGroupRelationState } from './memberGroupRelation.state';
+
```
Comment (naturetime10) [inline src/storage/states/app.state.ts:26]:
```suggestion
```

File: src/storage/states/versionSelector.state.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    public static fromJson(json: any): ExperimentVersionSelectorState {
+        return new ExperimentVersionSelectorState(
+            json.versionNumbers,
+        );
+    }
+}
+
+
```
Comment (naturetime10) [inline src/storage/states/versionSelector.state.ts]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/entity/group.ts [hunk 7]
```diff
@@ -1,23 +1,32 @@
...
 
-export type Group<GroupMember> =
-    | StaticGroup<GroupMember>
-    | FilterGroup<GroupMember>;
 
-export interface StaticGroup<GroupMember> {
+export type Group = StaticGroup<User> | StaticGroup<Team> | FilterGroup;
+
+export type GroupType = 'STATIC' | 'FILTER'
+
```
Comment (naturetime10) [inline src/entity/group.ts:10]:
```suggestion
```

File: src/entity/group.ts [hunk 8]
```diff
@@ -1,23 +1,32 @@
...
-    type: 'filter';
-    name: string;
-    filter: string;
-    createdAt: Date;
-    rollouts: Rollout<GroupMember>[];
+export interface StaticGroup<Member> extends GroupInfo {
+    type: 'STATIC';
+    members: Member[];
 }
+
```
Comment (naturetime10) [inline src/entity/group.ts:25]:
```suggestion
```

File: src/entity/rollout.ts [hunk 9]
```diff
@@ -1,68 +1,60 @@
...
     updatedAt?: Date;
 }
 
-export interface TimeRangeActivator {
-    id: number;
-    type: 'timeRange';
+export interface StaticActivator extends ActivatorInfo {
+    type: 'STATIC';
+}
+export interface TimeRangeActivator extends ActivatorInfo{
```
Comment (naturetime10) [inline src/entity/rollout.ts]:
```suggestion
}

export interface TimeRangeActivator extends ActivatorInfo{
```

File: src/entity/rollout.ts [hunk 10]
```diff
@@ -1,68 +1,60 @@
...
 
 export type VersionSelector = StaticVersionSelector | ExperimentVersionSelector;
 
-export interface StaticVersionSelector {
+
+export interface VersionSelectorInfo {
     id: number;
-    type: 'static';
-    versionNumber: number;
+    type: 'STATIC' | 'EXPERIMENT';
```
Comment (naturetime10) [inline src/entity/rollout.ts]: VersionSelectorType

File: src/storage/graph/activator.node.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+    get createdAt(): Date {
+        return this.activator.createdAt;
+    }
+
+    get updatedAt(): Date | undefined {
+        return this.activator.updatedAt;
+    }
+}
+
+
```
Comment (naturetime10) [inline src/storage/graph/activator.node.ts]:
```suggestion
```

File: src/storage/graph/activator.node.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+    constructor(
+        protected localStore: LocalStore,
+        protected activator: ActivatorState,
+    ) {}
+
+    get id(): number {
+        return this.activator.id;
+    }
+
+    get type(): 'STATIC' | 'TIME_RANGE' | 'MAX_VIEWERS' | 'PERCENTAGE' {
```
Comment (naturetime10) [inline src/storage/graph/activator.node.ts]: Can we define type in entity?

File: src/storage/graph/app.node.ts [hunk 13]
```diff
@@ -64,6 +69,67 @@ export class AppNode implements App {
...
+
+    public get teamGroups(): (StaticGroup<TeamNode> | FilterGroupNode)[] {
+        const appState = this.localStore.getState();
+        const app = appState.apps[this.app.id]; 
+        return app.teamGroupIds.map((groupId) => {
+            const group =  appState.groups[groupId];
+
```
Comment (naturetime10) [inline src/storage/graph/app.node.ts]:
```suggestion
```

File: src/storage/graph/group.node.ts [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+    constructor(
+        protected localStore: LocalStore,
+        protected group: GroupState,
+    ) {}
+
+    public get id(): number {
+        return this.group.id;
+    }
+
+    public get type(): 'STATIC' | 'FILTER' {
```
Comment (naturetime10) [inline src/storage/graph/group.node.ts]: Can we define type in entity?

File: src/storage/graph/group.node.ts [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+        return this.group.createdAt;
+    }
+
+    public get updatedAt(): Date | undefined {
+        return this.group.updatedAt;
+    }
+
+    public get rollouts(): RolloutNode[] {
+        const appState = this.localStore.getState();
+
```
Comment (naturetime10) [inline src/storage/graph/group.node.ts]:
```suggestion
```

File: src/storage/graph/rollout.node.ts [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+    get updatedAt(): Date | undefined {
+        return this.rollout.updatedAt;
+    }
+
+    get activator():
+        | StaticActivatorNode
+        | TimeRangeActivatorNode
+        | MaxViewersActivatorNode
+        | PercentageActivatorNode {
```
Comment (naturetime10) [inline src/storage/graph/rollout.node.ts]: Can we have a union type called `ActivatorNode`?

File: src/storage/graph/versionSelector.node.ts [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    constructor(
+        protected localStore: LocalStore,
+        protected versionSelector: VersionSelectorState,
+    ) {}
+
+    get id(): number {
+        return this.versionSelector.id;
+    }
+
+    get type(): 'STATIC' | 'EXPERIMENT' {
```
Comment (naturetime10) [inline src/storage/graph/versionSelector.node.ts]: Can we define versionSelectorType in entity?

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/storage/states/group.state.ts [hunk 18]
```diff
@@ -0,0 +1,58 @@
+export type GroupType = 'STATIC' | 'FILTER';
+
+export class GroupState {
+    constructor(
+    public id: number,
+    public type: GroupType,
+    public name: string,
+    public createdAt: Date,
+    public rolloutIds: number[],
```
Comment (naturetime10) [inline src/storage/states/group.state.ts:9]: Can we create rolloutGroupRelationState to support querying from group to rollout and rollout to group?

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/entity/rollout.ts [hunk 19]
```diff
@@ -1,68 +1,64 @@
...
-export interface Rollout<GroupMember> {
+export interface Rollout {
     id: number;
-    name: string;
     activator: Activator;
     versionSelector: VersionSelector;
     isEnabled: boolean;
     createdAt: Date;
     updatedAt?: Date;
-    groups: Group<GroupMember>[];
```
Comment (naturetime10) [inline src/entity/rollout.ts]: Can we add name and groups.

File: src/storage/graph/app.node.ts [hunk 20]
```diff
@@ -64,6 +68,65 @@ export class AppNode implements App {
         });
     }
 
+
```
Comment (naturetime10) [inline src/storage/graph/app.node.ts]:
```suggestion
```

File: src/storage/states/versionSelector.state.ts [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+    public static fromJson(json: any): VersionSelectorState {
+        return new VersionSelectorState(
+            json.id,
+            json.type,
+            new Date(json.createdAt),
+            new Date(json.updatedAt),
+        );
+    }
+}
+
```
Comment (naturetime10) [inline src/storage/states/versionSelector.state.ts:29]:
```suggestion
```

File: src/storage/states/versionSelector.state.ts [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+        );
+    }
+
+    public static fromJson(json: any): StaticVersionSelectorState {
+        return new StaticVersionSelectorState(
+            json.versionNumber,
+        );
+    }
+}
+
```
Comment (naturetime10) [inline src/storage/states/versionSelector.state.ts:47]:
```suggestion
```

**Review by naturetime10 — CHANGES_REQUESTED** [review]

File: src/entity/rollout.ts (hunk 19 above)
Comment (duanxianbo) [reply src/entity/rollout.ts]: added

File: src/storage/states/group.state.ts (hunk 18 above)
Comment (duanxianbo) [reply src/storage/states/group.state.ts:9]: will do in next pr

File: src/storage/graph/rollout.node.ts [hunk 23]
```diff
@@ -65,4 +70,11 @@ export class RolloutNode implements Rollout {
                 return new ExperimentVersionSelectorNode(this.localStore, versionSelector, experimentVersionSelector);
         }
     }
+
```
Comment (naturetime10) [inline src/storage/graph/rollout.node.ts:73]:
```suggestion

```

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**